packages/client/tests/functional/methods/upsert/native-atomic/tests.ts

```diff
@@ -0,0 +1,241 @@
+
+      const sqlPromise: Promise<string> = createSQLPromise(client)
+      let doesSQLContainOnConflict = false
+
+      // This 'will' use ON CONFLICT
+      await client.user.upsert({
+        where: {
+          // Because only one unique field is used
+          name,
+        },
+        create: {
+          name,
+        },
+        update: {
+          name,
+        },
+      })
+      doesSQLContainOnConflict = (await sqlPromise).includes('ON CONFLICT')
+      expect(doesSQLContainOnConflict).toEqual(true)
+    })
+
+    test('should only use ON CONFLICT when the unique field defined in where clause has the same value as defined in the create arguments', async () => {
+      const name = faker.name.firstName()
+      const title = faker.name.jobTitle()
+
+      await client.user.create({
+        data: {
+          name,
+          posts: {
+            create: {
+              title,
+            },
+          },
+        },
+      })
+
+      let sqlPromise: Promise<string> = createSQLPromise(client)
+      let doesSQLContainOnConflict = false
+
+      // This will 'not' use ON CONFLICT
+      await client.user.upsert({
+        where: {
+          name,
+        },
+        create: {
+          // Because the the where 'name' is 'not' equal to the create 'name'
+          name: name + '1',
+        },
+        update: {
+          name: name + '1',
+        },
+      })
+      doesSQLContainOnConflict = (await sqlPromise).includes('ON CONFLICT')
+      expect(doesSQLContainOnConflict).toEqual(false)
+
+      sqlPromise = createSQLPromise(client)
+      // This 'will' use ON CONFLICT
+      await client.user.upsert({
+        where: {
+          name,
+        },
+        create: {
+          // Because the the where 'name' is 'equal' to the create 'name'
+          name,
+        },
+        update: {
+          name: name + '1',
+        },
+      })
+      doesSQLContainOnConflict = (await sqlPromise).includes('ON CONFLICT')
+      expect(doesSQLContainOnConflict).toEqual(true)
+    })
+
+    test('should perform an upsert using ON CONFLICT', async () => {
+      const name = faker.name.firstName()
+      const title = faker.name.jobTitle()
+
+      await client.user.create({
+        data: {
+          name,
+          posts: {
+            create: {
+              title,
+            },
+          },
+        },
+      })
+
+      const sqlPromise: Promise<string> = createSQLPromise(client)
+      let doesSQLContainOnConflict = false
+
+      await client.user.upsert({
+        where: {
+          name,
+        },
+        create: {
+          name,
+        },
+        update: {
+          name,
+        },
+      })
+      doesSQLContainOnConflict = (await sqlPromise).includes('ON CONFLICT')
+      expect(doesSQLContainOnConflict).toEqual(true)
+    })
+  },
+  {
+    optOut: {
+      from: ['mongodb', 'mysql', 'sqlserver'],
+      reason: 'Other providers do not support native INSERT ... ON CONFLICT SET .. WHERE',
```

Comment:
Do we have normal tests for these other providers that just make sure that `upsert` works as it should?
Might it be worth adding test functionality here as well that tests the actual "update (if exists already) or insert (if not)" functionality of these upserts?

Reply:
Added simple upsert test's